freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRevertHandler.java

```diff
@@ -0,0 +1,58 @@
+package com.sequenceiq.freeipa.flow.stack.upgrade.ccm.handler;
+
+import static com.sequenceiq.freeipa.flow.stack.upgrade.ccm.selector.UpgradeCcmHandlerSelector.UPGRADE_CCM_REVERT_FAILURE_EVENT;
+import static com.sequenceiq.freeipa.flow.stack.upgrade.ccm.selector.UpgradeCcmStateSelector.UPGRADE_CCM_FAILED_EVENT;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.UpgradeCcmService;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmFailureEvent;
+
+import reactor.bus.Event;
+
+@Component
+public class UpgradeCcmRevertHandler extends ExceptionCatcherEventHandler<UpgradeCcmFailureEvent> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UpgradeCcmRevertHandler.class);
+
+    @Inject
+    private UpgradeCcmService upgradeCcmService;
+
+    @Override
+    public String selector() {
+        return UPGRADE_CCM_REVERT_FAILURE_EVENT.event();
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmFailureEvent> event) {
+        LOGGER.error("Reverting back has been failed", e);
+        UpgradeCcmFailureEvent resultEvent = UPGRADE_CCM_FAILED_EVENT.createBasedOn(event.getData());
+        resultEvent.setStatusReason(event.getData().getStatusReason() + " Reverting back has been failed");
+
+        return resultEvent;
+    }
+
+    @Override
+    protected Selectable doAccept(HandlerEvent<UpgradeCcmFailureEvent> event) {
+        UpgradeCcmFailureEvent request = event.getData();
+        LOGGER.info("CCM upgrade reverting");
+        try {
+            upgradeCcmService.changeTunnel(request.getResourceId(), request.getOldTunnel());
+            upgradeCcmService.pushSaltStates(request.getResourceId());
+        } catch (Exception e) {
+            LOGGER.error("Exception happened when reverting ccm upgrade. Flow finished ", e);
+        }
+        UpgradeCcmFailureEvent resultEvent = UPGRADE_CCM_FAILED_EVENT.createBasedOn(request);
+        resultEvent.setStatusReason(request.getStatusReason() + " Reverting back has been failed");
```

Comment:
same here, I would expect this would mean a successful revert and in the catch part a different event/message

Reply:
whether the revert success or not, the flow is failed - the ccm upgrade was unsuccessful